### PR TITLE
Resolve #157 - Create sort order for sidebar

### LINKED ISSUE #157: Order categories
I think we should order them differently:
- General usage
- Offline
- More than offline
- Web Push

...Instead of:
- General usage
- More than offline
- Offline
- Web Push

### PR DESCRIPTION
**Pull request by darkwing**
I also changed "More than Offline" to "Beyond Offline"

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -23,6 +23,8 @@ var srcRecipes = recipeSlugs.map(function makePath(name) {
   return './' + name + '/**';
 });
 
+var categoryOrder = ['General Usage', 'Offline', 'Beyond Offline', 'Performance', 'Web Push'];
```
Comment (marco-c) [inline gulpfile.js]: Could you add an assertion to check that a recipe belongs to one of these categories?

**Comment by darkwing** [comment]
Updated!  Let me know if it's to your liking!

File: parseRecipes.js [hunk 2]
```diff
@@ -27,13 +35,10 @@ module.exports = function(recipeSlugs) {
...
-      difficulty = 3;
-    } else if (difficultyTerm === 'Intermediate') {
-      difficulty = 2;
-    } else if (difficultyTerm === 'Beginner') {
-      difficulty = 1;
-    } else {
+    if(difficultyTerm in difficulties) {
+      difficulty = difficulties[difficultyTerm];
+    }
+    else {
```
Comment (marco-c) [inline parseRecipes.js]: Huge nit: } else {

**Comment by marco-c** [comment]
> Updated! Let me know if it's to your liking!

I like that you've moved everything in the parseRecipes function! You haven't added the assertion though :smiley:

**Comment by darkwing** [comment]
Doh, had un-pushed commit.  Updated!

File: parseRecipes.js [hunk 3]
```diff
@@ -58,5 +62,19 @@ module.exports = function(recipeSlugs) {
...
+    }
+
+    if (recipeA.category < recipeB.category) {
+      return -1;
+    }
+
+    return 1;
+  }).sort(function(recipeA, recipeB) {
+    // Sort recipes by preferred category
+    assert(categoryOrder.indexOf(recipeA.category) !== -1, recipeA.name + ': Unexpected category value "' + recipeA.category + '"');
```
Comment (marco-c) [inline parseRecipes.js]: With `sort` some elements of the array could be missed. E.g. if you run this code:
```
[3, 4, 5].sort(function(a, b) {
  console.log(a);
  return a - b;
}
```
only *3* and *4* will be logged.

`forEach` would definitely be called for all the elements.

Comment (darkwing) [inline parseRecipes.js]: Wow, I didn't know that; how can we use `forEach` to do the sort?

Comment (marco-c) [inline parseRecipes.js]: You can use `sort` to sort and then `forEach` to assert (or the other way around).

Comment (darkwing) [inline parseRecipes.js]: Ohh, your concern is the assert and not the correct sort?  OK, that makes sense.  Will update.

**Comment by darkwing** [comment]
Grrr, the duplicate sort here either orders the categories wrong or the difficulties wrong, based on the order of the sort calls; this will require more changes.

**Comment by marco-c** [comment]
Nit: I was thinking that instead of adding the `assert` with the `forEach`, you could add it in the mapping function (`recipeSlugs.map(function(recipe) {`), where other assertions are.

To fix the sorting problem, you can put everything in a single sorting function, something like:
```JavaScript
function(recipeA, recipeB) {
  // Sort recipes by preferred category
  if (categoryOrder.indexOf(recipeA.category) === categoryOrder.indexOf(recipeB.category)) {
    // Sort by difficulty
  } else if (categoryOrder.indexOf(recipeA.category) > categoryOrder.indexOf(recipeB.category)) {
    return 1;
  } else {
    return -1;
  }
}
```

**Comment by darkwing** [comment]
Ahh, very nice.  Updated!

**Pull request merged**